### PR TITLE
Refactoring

### LINKED ISSUE #12: wip if-watch rewrite
so the current if-watch needs some maintenance. currently working on adding support for the remaining platforms (macos, ios, android) and doing some refactoring. on linux we should use the rtnetlink crate and on windows the official (from microsoft) windows crate. on android we no longer have access to rtnetlink (due to a fairly recent selinux policy update) so we need to bind to the java api to receive notifications.

### PR DESCRIPTION
**Pull request by dvc94ch**
TODO:
- [ ] android implementation
- [x] wait for new release of system-configuration (https://github.com/mullvad/system-configuration-rs/pull/30)

Closes #12 

So decided not to implement it on android for now. it would require some java glue and javac to build the crate. the fallback implementation should work on android for now.

### CODE REVIEW

**Comment by dvc94ch** [comment]
ping @mxinden

**Comment by dvc94ch** [comment]
the api hasn't changed, so this is a minor version bump

**Comment by dvc94ch** [comment]
ping @mxinden

File: Cargo.toml [hunk 1]
```diff
@@ -8,21 +8,33 @@ license = "MIT OR Apache-2.0"
 description = "crossplatform asynchronous network watcher"
 repository = "https://github.com/dvc94ch/if-watch"
 
+[features]
+force-fallback = ["async-io", "if-addrs"]
```
Comment (mxinden) [inline Cargo.toml:12]: I am not quite sure I understand why the `fallback` feature is needed. In which case would one want to force a _fallback_ even though the used platform is directly supported? Would you mind documenting this?

File: src/linux.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    }
+}
+
+impl IfWatcher {
+    pub async fn new() -> Result<Self> {
+        let (mut conn, handle, messages) = rtnetlink::new_connection_with_socket::<SmolSocket>()?;
+        let groups = RTMGRP_IPV4_IFADDR | RTMGRP_IPV6_IFADDR;
+        let addr = SocketAddr::new(0, groups);
+        conn.socket_mut().socket_mut().bind(&addr)?;
+        let task = async_global_executor::spawn(conn);
```
Comment (mxinden) [inline src/linux.rs:39]: Stupid question: Given that `IfWatch` has a `Future::poll` method, why does it need to be spawned onto an executor?

**Review by mxinden — COMMENTED** [review]
Quite a large change. Would you mind enumerating the various changes to ease the review?

File: Cargo.toml (hunk 1 above)
Comment (dvc94ch) [reply Cargo.toml:12]: Mainly so you can test that the fallback works. Now you can compile for Android to test the fallback, but was running out of targets to compile for. The fallback doesn't work on wasm32 or wasi

File: src/linux.rs (hunk 2 above)
Comment (dvc94ch) [reply src/linux.rs]: had some trouble getting it to work first time around. was no trouble this time.

**Comment by dvc94ch** [comment]
The changes were:

- [ ] rewrite linux backend using rtnetlink
- [ ] add new macos/ios backend
- [ ] use windows crate instead of winapi crate in windows backend

**Review by mxinden — APPROVED** [review]
Thanks for removing the extra `spawn` call!

> the api hasn't changed, so this is a patch version bump. we can also make it 1.0.0

I am fine either way. How about a basic changelog e.g. via [keep a changelog](https://keepachangelog.com/)? That would help downstream users to understand the new changes.

**Pull request merged**